engine/gamesys/src/gamesys/test/test_gamesys.h

```diff
@@ -39,8 +38,38 @@ struct Params
 };
 
 template<typename T>
-class GamesysTest : public jc_test_params_class<T>
+class CustomizableGamesysTest : public jc_test_params_class<T>
+{
+public:
+    struct ProjectOptions {
+      uint32_t m_MaxCollisionCount;
+      uint32_t m_MaxContactPointCount;
+      bool m_3D;
+    };
+
+    CustomizableGamesysTest() {
+      memset(&projectOptions, 0, sizeof(projectOptions));
+    }
+
+    ProjectOptions projectOptions;
+protected:
+    void SetUp(ProjectOptions& options) {
+      projectOptions = options;
+    }
+
+    virtual void SetUp() = 0;
+};
+
+template<typename T>
+class GamesysTest : public CustomizableGamesysTest<T>
 {
+public:
+    GamesysTest() {
+        // Default configuration values for the engine. Subclass GamesysTest and ovewrite in constructor.
+        this -> projectOptions.m_MaxCollisionCount = 0;
```

Comment:
We try to keep the code style of the existing files.
For instance, here we don't use "this".
Also, if we were using it, it would be "this->" (without the spaces)

Reply:
Ok, let me check again about the use of `this` because i thing there is a hard-to-explain ambiguity with the C++ template there. I didn't have it in the beginning and put it afterwards to get things right.

Reply:
@JCash , i did some research and experimented too. The bottom line is that when there is a child class template that tries to access of a member of a parent class template things get complicated. 

Note, i'm  talking about _child class templates_ (like `template<typename T> class GamesysTest : public CustomizableGamesysTest<T>`) which are still generic, and not _child template classes_ (like `class Sleeping2DCollisionObjectTest : public GamesysTest<const char*>`) which are more concrete. This behavior was not apparent so far because access to `GamesysTest` members was done by concrete _child template classes_.

The best discussion i could find online is this:

https://stackoverflow.com/questions/1120833/derived-template-class-access-to-base-class-member-data  

#### Solutions

1. Use `this`. With spaces i promise :-)
2. Use more explicit member declaration  

```
GamesysTest() {
    // Default configuration values for the engine. Subclass GamesysTest and ovewrite in constructor.
    CustomizableGamesysTest<T>::m_projectOptions.m_MaxCollisionCount = 0;
```

3. Use the 'using' keyword in the functions where m_projectOptions is accessed. Note, it didn't work inside constructors. Not sure why. The following worked:

```
template<typename T>
class GamesysTest : public CustomizableGamesysTest<T>
{
    ...
    void testMemberAccess() {
          using CustomizableGamesysTest<T>::m_projectOptions;
          m_projectOptions.m_3D = false;
    }
```

I would vote 1 or 2.

Reply:
I still don't get it. It didn't compile if you removed the "this->" part?

Reply:
@JCash , here is our case...

I compiled this:
```
template<typename T>
class GamesysTest : public CustomizableGamesysTest<T>
{
public:
    GamesysTest() {
        // Default configuration values for the engine. Subclass GamesysTest and ovewrite in constructor.
        m_projectOptions.m_MaxCollisionCount = 0;
        this->m_projectOptions.m_MaxContactPointCount = 0;
        this->m_projectOptions.m_3D = false;
    }
```

and got this:
```
[393/396] cxx: src/gamesys/test/test_gamesys.cpp -> build/default/src/gamesys/test/test_gamesys_1.o                                                                                          
13:54:01 runner system command -> ['/usr/bin/ccache', '/home/nando/src/defold/tmp/dynamo_home/ext/SDKs/linux/clang-9.0.0/bin/clang++', '-O0', '-std=c++11', '-g', '-D__STDC_LIMIT_MACROS', '-DDDF_EXPOSE_DESCRIPTORS', '-DGOOGLE_PROTOBUF_NO_RTTI', '-Wall', '-Werror=format', '-fno-exceptions', '-fPIC', '-fvisibility=hidden', '-fno-rtti', '-DDM_PLATFORM_64BIT', '-Idefault/src', '-I../src', '-Idefault/proto', '-I../proto', '-I/home/nando/src/defold/tmp/dynamo_home/ext/include', '-I/home/nando/src/defold/tmp/dynamo_home/sdk/include', '-I/home/nando/src/defold/tmp/dynamo_home/include', '-I/home/nando/src/defold/tmp/dynamo_home/include/x86_64-linux', '-I/home/nando/src/defold/tmp/dynamo_home/ext/include/x86_64-linux', '-DDLIB_LOG_DOMAIN="GAMESYS"', '-DLUA_BYTECODE_ENABLE_64', '-DLUA_BYTECODE_ENABLE_64', '../src/gamesys/test/test_gamesys.cpp', '-c', '-o', 'default/src/gamesys/test/test_gamesys_1.o']
In file included from ../src/gamesys/test/test_gamesys.cpp:13:
../src/gamesys/test/test_gamesys.h:69:9: error: use of undeclared identifier 'm_projectOptions'
        m_projectOptions.m_MaxCollisionCount = 0;
        ^
1 error generated.                                                                                                                                                                           
Waf: Leaving directory `/home/nando/src/defold/engine/gamesys/build'
Build failed
 -> task failed (err #1):                                                                                                                                                                    
        {task: cxx test_gamesys.cpp -> test_gamesys_1.o}   
``